### PR TITLE
Add prepared statements support for `Mysql2Adapter`

### PR DESCRIPTION
**Pull request by kamipo**

### CODE REVIEW

**Comment by schneems** [comment]
r? @sgrif

We probably need your mysql2 branch to be merged into the official mysql2 gem before moving forwards

**Comment by sgrif** [comment]
Yeah, this isn't actionable as is. I'm going to close just because there's 0 chance of this being merged right now. Please reopen one you've merged into MySQL 2

File: activerecord/lib/active_record/connection_adapters/mysql_adapter.rb [hunk 1]
```diff
@@ -317,41 +308,10 @@ def initialize_type_map(m) # :nodoc:
...
-
-          [result_set, affected_rows]
-        end
-      end
-
       def execute_and_free(sql, name = nil) # :nodoc:
         result = execute(sql, name)
         ret = yield result
-        result.free
+        result.free if result
```
Comment (sodabrew) [inline activerecord/lib/active_record/connection_adapters/mysql_adapter.rb]: @kamipo Do you specifically need mysql2 to provide Mysql2::Result#free, or would it be sufficient to allow Ruby GC to take care of the result set? I sort of imagine that this method was written for the original Mysql gem that required the explicit free.

That said, if it helps to land this into Rails sooner, then I'll be sure to provide a Mysql2::Result#free per your PR.

**Comment by sodabrew** [comment]
I'm planning to have the next mysql2 gem released before US Thanksgiving next week, so if there's any possibility of incorporating this into the next Rails 4.2.5 release, that would be very exciting.

**Comment by sodabrew** [comment]
Here we go! mysql2 0.4.2 is released! Please let me know if I can help this PR in any way!

**Comment by sgrif** [comment]
@kamipo If you want to update and open a new PR, that'd be great.

@sodabrew This won't make it into 4.2.x, we don't add new features in patch releases.

**Comment by kamipo** [comment]
Reopened at #22415.

**Pull request closed without merging**